src/SherlogN.jl

```diff
@@ -0,0 +1,95 @@
+Base.Int32(x::$SherlogN) = Int32(x.val)
+Base.Int16(x::$SherlogN) = Int16(x.val)
+
+# otherwise conversion to itself is ambigious
+$SherlogN{T,i}(x::$SherlogN) where {T<:AbstractFloat,i} = $SherlogN{T,i}(x.val)
+$SherlogN(x::$SherlogN{T,i}) where {T<:AbstractFloat,i} = $SherlogN{T,i}(x.val)
+
+# generator functions from Float, Int and Bool
+$SherlogN(x::AbstractFloat) = $SherlogN{Float16,1}(x)   # use Float16 and logbook #1 as standard
+$SherlogN{T}(x::AbstractFloat) where T = $SherlogN{T,1}(x)      # use logbook #1 if no logbook provided
+$SherlogN(x::Integer) = $SherlogN{Float16,1}(x)           # same for integers
+$SherlogN{T}(x::Integer) where T = $SherlogN{T,1}(x)
+$SherlogN{T}(x::Bool) where T = if x $SherlogN{T,1}(1) else $SherlogN{T,1}(0) end
+
+Base.promote_rule(::Type{<:Integer},::Type{$SherlogN{T,i}}) where {T,i} = $SherlogN{T,i}
+Base.promote_rule(::Type{Float64},::Type{$SherlogN{T,i}}) where {T,i} = $SherlogN{T,i}
+Base.promote_rule(::Type{Float32},::Type{$SherlogN{T,i}}) where {T,i} = $SherlogN{T,i}
+Base.promote_rule(::Type{Float16},::Type{$SherlogN{T,i}}) where {T,i} = $SherlogN{T,i}
+
+Base.bitstring(x::$SherlogN) = bitstring(x.val)
+Base.show(io::IO,x::$SherlogN) = print(io,$SherlogN,"(",string(x.val),")")
+
+Base.eps(::Type{$SherlogN{T,i}}) where {T,i} = eps(Float64)
+Base.eps(x::$SherlogN) = eps(x.val)
+
+Base.typemin(::$SherlogN{T,i}) where {T,i} = $SherlogN{T,i}(typemin($Float))
+Base.typemax(::$SherlogN{T,i}) where {T,i} = $SherlogN{T,i}(typemax($Float))
+Base.floatmin(::$SherlogN{T,i}) where {T,i} = $SherlogN{T,i}(floatmin($Float))
+Base.floatmax(::$SherlogN{T,i}) where {T,i} = $SherlogN{T,i}(floatmax($Float))
+Base.precision(::$SherlogN{T,i}) where {T,i} = $SherlogN{T,i}(precision($Float))
+
+end
+
+for O in (:(+), :(-), :(*), :(/), :(^))
+    @eval function Base.$O(x::$SherlogN{T,i},y::$SherlogN{T,i}) where {T,i}
+        r = $O(x.val, y.val)
+        log_it(T,r,i)
+        return $SherlogN{T,i}(r)
+    end
+end
+
+for O in (:(-), :(+),
+          :sign,
+          :prevfloat, :nextfloat,
+          :round, :trunc, :ceil, :floor,
+          :inv, :abs, :sqrt, :cbrt,
+          :exp, :expm1, :exp2, :exp10,
+          :log, :log1p, :log2, :log10,
+          :rad2deg, :deg2rad, :mod2pi, :rem2pi,
+          :sin, :cos, :tan, :csc, :sec, :cot,
+          :asin, :acos, :atan, :acsc, :asec, :acot,
+          :sinh, :cosh, :tanh, :csch, :sech, :coth,
+          :asinh, :acosh, :atanh, :acsch, :asech, :acoth,
+          :sinc, :sinpi, :cospi,
+          :sind, :cosd, :tand, :cscd, :secd, :cotd,
+          :asind, :acosd, :atand, :acscd, :asecd, :acotd
+         )
+    @eval function Base.$O(x::$SherlogN{T,i}) where {T,i}
+        r = $O(x.val)
+        log_it(T,r,i)
+        return Sherlog64{T,i}(r)
+    end
+end
+
+for O in (:(<), :(<=))
+    @eval Base.$O(x::$SherlogN, y::$SherlogN) = $O(x.val, y.val)
+end
+
+@eval function rand!(x::Array{$SherlogN})
+    for i in eachindex(x)
+        x[i] = $SherlogN.(rand($Float))
```

Comment:
Can we write this as `rand($SherlogN)` and define that function as something like `rand(::Type{T}) where {T<:AbstractSherlog} = convert(T,rand($Float))` ? And maybe while we're at it also define `randn`?